Issue #89 - add a command line flag to export for sanitize level

janetriley opened the pull request:
* Added -Z and --sanitize flag to command
 * Added sanitize_level attribute to Exporter class in the initializer, to be consistent with scheme
 * Addded sanitize_level argument to export method 
 * Added tests for invalid scheme and sanitize level
* '/tmp/corpus' was getting repeated a lot in tests, so set a self attribute in setUp.
* Restructured export method into smaller methods, since I was doing some duplicate work in __init__ and export, and it makes them easier to test. 
* Don't set object attributes in export(); make export related functions take values from parameters rather than self.  Previously export() updated the self values with the parameter values, and it feels weird to make a persisting change to the object when calling export.  Sending values in parameters means not having to look them up in self.

janetriley committed:
Issue #89 - add a command line flag to export for sanitize level

 * Added flag to command
 * Added sanitize_level attribute to Exporter model initializer
 * Addded sanitize_level argument to export method
 * Restructured export method into smaller methods, make them take values from parameters rather than self
 * Added tests for invalid scheme and sanitize level

bbengfort commented on baleen/export.py:
```diff
@@ -169,60 +179,98 @@ def feedinfo(self, path):
+            raise ExportError(INVALID_SANITIZE_LEVEL.format(level))
+
+        self.initialize_export_directory(root)
+        category_filepaths = self.initialize_category_dirs(base_dir=root,    
```
From what I understand - this will create all the subdirectories in the exporter. If there are no posts with an associated category, this means that an empty directory will be created on export, right?

janetriley replied:
Yes, it looks like that.  Take 2 is in the revised version.

bbengfort commented on baleen/export.py:
```diff
@@ -169,60 +179,98 @@ def feedinfo(self, path):
-            os.mkdir(self.root)
+        # Iterate through all posts, writing them to disk correctly.
+        # Right now we will simply write them based on their object id.
+        for post, category in tqdm(self.posts(categories=categories),
```
What is `tqdm`? I don't see it defined, but it could be that the code review is collapsing code.

janetriley replied:
[Tqdm wraps iterators in a progress bar](https://pypi.python.org/pypi/tqdm).   That's as I found it.

bbengfort replied:
Hehe ... well I should have known that; the problem was that I was looking in the master branch not the develop branch and didn't see the import -- it's all good!

bbengfort commented on tests/test_export.py:
```diff
@@ -103,6 +103,8 @@ def setUpClass(cls):
         assert Feed.objects.count() == 3
         assert Post.objects.count() == 3
+        cls.root_dir = "/tmp"
+        cls.corpus_dir = "/tmp/corpus"
```
I know this hard code is inherited, but while I'm reviewing I'd suggest that we use [`mkdtemp()`](https://docs.python.org/3.5/library/tempfile.html#tempfile.mkdtemp) for tests and ensure that we also clean up the tests by deleting the directory when create when we're done testing.

janetriley replied:
Et voila. 

I got errors about lambdas when I handled it in the  setUp and tearDown, so I cleaned up the corpus directory on tearDown, and root_dir in class teardown.

bbengfort reviewed (approved):
This looks great; I reviewed the code line by line and didn't see anything that jumped out at me. Everything looks very neat and well composed. Nice work!

janetriley committed:
Replace hardcoded /tmp with mkdtemp() dir

janetriley committed:
Issue #89 - add a command line flag to export for sanitize level

Use os.path.join for test directory paths.

janetriley committed:
Issue #89 - add a command line flag to export for sanitize level

Manage our directories better:  use TemporaryDirectory, clean up in tearDown.

janetriley committed:
Issue #89 - add a command line flag to export for sanitize level

Create category directories iff there are export results rather than one per all categories.

janetriley committed:
Fix failing test: mock the post method for missing mongo rewind()

janetriley committed:
Make export() raise ExportError on file exception, to match test expectation

janetriley commented:
The advantage of creating all category directories at the start is it raises an exception immediately if there's an issue creating the directory, rather than throwing it in the midst of an export.  That seems like a worthwhile tradeoff once you have a big corpus.

What do you think, @bbengfort , back it out?

I'm not thrilled with raising the ExportException for FileNotFound in the midst of the loop. It seems odd to single that one exception type out.  I can remove it, reraise any exception as an ExportError, or leave be.

bbengfort commented:
I'd say go ahead and create all the directories at the beginning - an empty directory is no big deal, and I think you're right that we don't want to throw an exception in the middle of the export. 

Which brings us to the exceptions that are thrown in the middle of the export. A couple of points:

Any exception that is not of type `BaleenError` should be considered a fatal developer error. I think that's why the `FileNotFound` exception is being caught and wrapped in `ExportError` because any other type of IOError should simply cause the program to crash and the developers have to deal with it as a bug (which you expressed in your point about so many ways for I/O to go wrong). 

At this point though Baleen should be mature enough to gracefully handle exceptions, we've been running it long enough and haven't observed too much craziness. 

For now, I would say that we should catch and reraise any exception as an ExportError, since they are being handled in a specific way and require the errno and other fields. 

In the future, we could collect all errors in exporter.errors and show them to the user after export (or at least classify non fatal vs fatal export exceptions)

bbengfort commented:
Speaking of specific error handling and fatal vs. non fatal - most of that happens at the logger level.
